[16.0] [FIX] l10n_it_split_payment: impedire conferma account.move sbilanciate

patrickt-oforce opened the pull request:
Come indicato in #4832 quando una move viene duplicata, attulamente il modulo interferisce nella write impostando la chiave `check_move_validity` a False durante la chiamata alla super() e questo permette di confermare le account.move con righe sbilanciate

patrickt-oforce committed:
[FIX] l10n_it_split_payment: move set context flag to avoid confirm unbalanced account move

Applying context key `check_move_validity` as `False` on super() call, allow to validate a duplicated account.moe with unbalanced lines; moving the context set where needed avoid this behaviour

patrickt-oforce commented:
Test effettuato su runboat della localizzazione alla versione 16.0
<img width="3210" height="1524" alt="image" src="https://github.com/user-attachments/assets/d67b36d3-3cad-4106-835a-d5d3c042ba74" />

TheMule71 reviewed (changes requested):
@eLBati @matteoopenf
La modifica disfa parzialmente il codice precedente. Mi chiedo a cosa servisse disabilitare il controllo sulla write().

Io la correggerei cosi:
```
diff --git a/l10n_it_split_payment/models/account_move.py b/l10n_it_split_payment/models/account_move.py
index 654d65081..db82830c7 100644
--- a/l10n_it_split_payment/models/account_move.py
+++ b/l10n_it_split_payment/models/account_move.py
@@ -46,9 +46,8 @@ class AccountMove(models.Model):
         res = super(AccountMove, self.with_context(check_move_validity=False)).write(
             vals
         )
-        if self.env.context.get("skip_split_payment_computation"):
-            return res
-        self.compute_split_payment()
+        if not self.env.context.get("skip_split_payment_computation"):
+            self.compute_split_payment()
         container = {"records": self}
         self._check_balanced(container)
         return res
```

TheMule71 commented:
Per la cronaca, credo che il bug sia stato introdotto con https://github.com/OCA/l10n-italy/pull/4630

patrickt-oforce commented:
> Per la cronaca, credo che il bug sia stato introdotto con #4630

Si penso anche io sia entrato con quel commit, che ho solo riportato

patrickt-oforce commented:
> @eLBati @matteoopenf La modifica disfa parzialmente il codice precedente. Mi chiedo a cosa servisse disabilitare il controllo sulla write().
> 
> Io la correggerei cosi:
> 
> ```
> diff --git a/l10n_it_split_payment/models/account_move.py b/l10n_it_split_payment/models/account_move.py
> index 654d65081..db82830c7 100644
> --- a/l10n_it_split_payment/models/account_move.py
> +++ b/l10n_it_split_payment/models/account_move.py
> @@ -46,9 +46,8 @@ class AccountMove(models.Model):
>          res = super(AccountMove, self.with_context(check_move_validity=False)).write(
>              vals
>          )
> -        if self.env.context.get("skip_split_payment_computation"):
> -            return res
> -        self.compute_split_payment()
> +        if not self.env.context.get("skip_split_payment_computation"):
> +            self.compute_split_payment()
>          container = {"records": self}
>          self._check_balanced(container)
>          return res
> ```

Così credo che il problema non si risolva, perché nel context della super() viene inserito il controllo per evitare il controllo della validità della move, causa che permette di salvare la move sbilanciata.
@TheMule71 cosa intendi con disabilitare il controllo sulla write? e in che senso disfa il codice precedente?

TheMule71 commented:
`check_move_validity=False` è stata introdotta da @eLBati, immagino per un motivo. 
https://github.com/OCA/l10n-italy/pull/4630/commits/ab0c5358c8b34db028e07cf4a19e741cf4a9f7b2
se togli `check_move_validity=False`  probabilmente "rompi" qualsiasi cosa quella PR intendeva fixare.

Lì si faceva la write con  `check_move_validity=False` ma alla fine si chiamava la `_check_balanced()`, per evitare di creare move sbilanciate. Il problema sta nel `return res`.

Con la mia modifica qualunque percorso (`skip_split_payment_computation` che sia True o False), passa _sempre_ per la `_check_balanced()` quindi non può creare move sbilanciate. Risolve il problema senza rompere qualunque cosa la #4630 abbia fixato.

patrickt-oforce commented:
> `check_move_validity=False` è stata introdotta da @eLBati, immagino per un motivo. [ab0c535](https://github.com/OCA/l10n-italy/commit/ab0c5358c8b34db028e07cf4a19e741cf4a9f7b2) se togli `check_move_validity=False` probabilmente "rompi" qualsiasi cosa quella PR intendeva fixare.
> 
> Lì si faceva la write con `check_move_validity=False` ma alla fine si chiamava la `_check_balanced()`, per evitare di creare move sbilanciate. Il problema sta nel `return res`.
> 
> Con la mia modifica qualunque percorso (`skip_split_payment_computation` che sia True o False), passa _sempre_ per la `_check_balanced()` quindi non può creare move sbilanciate. Risolve il problema senza rompere qualunque cosa la #4630 abbia fixato.

Al netto che non l'ho tolto, ma messo nel context solo alla creazione della riga per lo split payment; quel flag serve appunto a garantire la creazione dell'account.move "sbilanciata" con una sola riga di imposte prima di fare tutti i suoi calcoli. Tenerlo nella super, porta alla condizione dello screen shot e che può essere testato su runbot della localizzazione italiana

TheMule71 commented:
> porta alla condizione dello screen shot e che può essere testato su runbot della localizzazione italiana

Anche con la mia modifica?

> ma messo nel context solo alla creazione della riga per lo split payment;

potrebbe essere troppo tardi.

eLBati commented:
@TheMule71 @patrickt-oforce l'obiettivo delle mie modifiche era risolvere questo https://github.com/OCA/l10n-italy/issues/4231

Quindi bisognerebbe accertarsi che quel problema non si ripresenti

patrickt-oforce commented:
> > porta alla condizione dello screen shot e che può essere testato su runbot della localizzazione italiana
> 
> Anche con la mia modifica?

Si, appena testato su un db demo, con solo lo split payment installato, e questo è il risultato
<img width="2468" height="1062" alt="image" src="https://github.com/user-attachments/assets/933d8c95-f1db-407a-82fc-4f6a6b3187b2" />


> 
> > ma messo nel context solo alla creazione della riga per lo split payment;
> 
> potrebbe essere troppo tardi.
Beh no, per la localizzazione è importante che faccia creare la sola riga di imposta dello split payment e poi Odoo faccia i dovuti calcoli per bilanciare il tutto

patrickt-oforce commented:
> @TheMule71 @patrickt-oforce l'obiettivo delle mie modifiche era risolvere questo #4231
> 
> Quindi bisognerebbe accertarsi che quel problema non si ripresenti

Nella issue / PR indicata, il problema era la perdita del giroconto dello split_payment nelle fatture, corretto?

TheMule71 commented:
Dalle prove che ho fatto (sul runboat di questa PR) non riesco a riprodurre il problema di Lorenzo, ma vedo questo stranamente:

<img width="1192" height="270" alt="image" src="https://github.com/user-attachments/assets/a6750794-2c15-4868-8b59-7f61323896a7" />

Appare una riga in più a zero (la seconda), dopo aver aggiornato il prezzo del prodotto.

patrickt-oforce committed:
[IMP] l10n_it_split_payment: avoid context key

Avoid using context key to force skip check on move validity,
instead add the value for tax_repartition_line_id in refund split
payment line to ensure the line is retained when the invoice is
confirmed

patrickt-oforce commented:
Sembra non essere più necessaria, procedo a chiuderla
